File: ui/src/assistantApi.ts
// Typed client for the backend assistant-management REST API
// (see stream-backend/app/assistant_api.py).
import { DEFAULT_API_URL } from "./api";

export type Permission = "allow" | "ask" | "deny";

export interface ModelConfig {
  provider: "google" | "anthropic" | "bedrock" | "openai";
  name: string;
  temperature: number;
  max_tokens: number | null;
  api_key?: string | null;
}

export interface ToolConfig {
  name: string;
  permission: Permission;
}

export interface MCPServerConfig {
  name: string;
  transport: "stdio" | "streamable_http" | "sse";
  command: string | null;
  args: string[];
  url: string | null;
  env: Record<string, string>;
  permission: Permission;
  enabled: boolean;
}

export interface SkillConfig {
  name: string;
  description: string;
  path: string;
  enabled: boolean;
}

export interface MemoryConfig {
  name: string;
  path: string;
  enabled: boolean;
}

export interface SubAgentConfig {
  name: string;
  description: string;
  system_prompt: string;
  tools: string[];
  skills: string[];
  model: string | null;
}

export interface MiddlewareConfig {
  name: string;
  enabled: boolean;
  config: Record<string, unknown>;
}

export interface AssistantConfig {
  assistant_id: string;
  name: string;
  description: string;
  created_at: string;
  updated_at: string;
  model: ModelConfig;
  system_prompt: string;
  tools: ToolConfig[];
  mcp: MCPServerConfig[];
  skills: SkillConfig[];
  memory: MemoryConfig[];
  subagents: SubAgentConfig[];
  middleware: MiddlewareConfig[];
  recursion_limit: number;
  metadata: Record<string, unknown>;
}

export interface CatalogTool {
  name: string;
  label?: string;
  description: string;
  kind: string;
}

export interface CatalogMiddleware {
  name: string;
  label: string;
  description: string;
  config_schema: Record<string, unknown>;
}

export interface CatalogModel {
  name: string;
  label: string;
}

export interface CatalogProvider {
  name: string;
  label: string;
  example: string;
  models?: CatalogModel[];
}

export interface Catalog {
  tools: CatalogTool[];
  builtin_tools: CatalogTool[];
  middleware: CatalogMiddleware[];
  providers: CatalogProvider[];
  permissions: Permission[];
}

export type AssistantUpsert = Omit<AssistantConfig, "assistant_id" | "created_at" | "updated_at"> & {
  assistant_id?: string;
};

async function json<T>(response: Response, action: string): Promise<T> {
  if (!response.ok) {
    let detail = response.statusText;
    try {
      const body = (await response.json()) as { detail?: string };
      if (body?.detail) detail = body.detail;
    } catch {
      /* ignore */
    }
    throw new Error(`${action} failed: ${detail}`);
  }
  return (await response.json()) as T;
}

export async function fetchCatalog(apiUrl = DEFAULT_API_URL): Promise<Catalog> {
  return json(await fetch(`${apiUrl}/assistants/catalog`), "Load catalog");
}

export async function listAssistants(apiUrl = DEFAULT_API_URL): Promise<AssistantConfig[]> {
  return json(await fetch(`${apiUrl}/assistants`), "List assistants");
}

export async function getAssistant(id: string, apiUrl = DEFAULT_API_URL): Promise<AssistantConfig> {
  return json(await fetch(`${apiUrl}/assistants/${id}`), "Get assistant");
}

export async function createAssistant(payload: AssistantUpsert, apiUrl = DEFAULT_API_URL): Promise<AssistantConfig> {
  return json(
    await fetch(`${apiUrl}/assistants`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }),
    "Create assistant",
  );
}

export async function updateAssistant(
  id: string,
  payload: AssistantUpsert,
  apiUrl = DEFAULT_API_URL,
): Promise<AssistantConfig> {
  return json(
    await fetch(`${apiUrl}/assistants/${id}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }),
    "Update assistant",
  );
}

export async function deleteAssistant(id: string, apiUrl = DEFAULT_API_URL): Promise<void> {
  const response = await fetch(`${apiUrl}/assistants/${id}`, { method: "DELETE" });
  if (!response.ok && response.status !== 204) {
    throw new Error(`Delete assistant failed: ${response.statusText}`);
  }
}

export async function writeSkill(
  id: string,
  body: { name: string; content: string; description?: string },
  apiUrl = DEFAULT_API_URL,
): Promise<SkillConfig> {
  return json(
    await fetch(`${apiUrl}/assistants/${id}/skills`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }),
    "Write skill",
  );
}

export async function writeMemory(
  id: string,
  body: { name: string; content: string },
  apiUrl = DEFAULT_API_URL,
): Promise<MemoryConfig> {
  return json(
    await fetch(`${apiUrl}/assistants/${id}/memory`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }),
    "Write memory",
  );
}

export interface AssistResult {
  content: string;
  source: "model" | "template";
}

export async function draftSystemPrompt(
  body: { name: string; description?: string; tools?: string[]; subagents?: string[]; instructions?: string; model?: ModelConfig },
  apiUrl = DEFAULT_API_URL,
): Promise<AssistResult> {
  return json(
    await fetch(`${apiUrl}/assistants/assist/system-prompt`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }),
    "Draft system prompt",
  );
}

export async function draftSkill(
  body: { name: string; description?: string; instructions?: string; model?: ModelConfig },
  apiUrl = DEFAULT_API_URL,
): Promise<AssistResult> {
  return json(
    await fetch(`${apiUrl}/assistants/assist/skill`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }),
    "Draft skill",
  );
}

export interface TestModelResult {
  ok: boolean;
  message: string;
  sample?: string;
}

export async function testModel(model: ModelConfig, apiUrl = DEFAULT_API_URL): Promise<TestModelResult> {
  return json(
    await fetch(`${apiUrl}/assistants/assist/test-model`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ model }),
    }),
    "Test model",
  );
}

export function emptyAssistant(): AssistantUpsert {
  return {
    name: "",
    description: "",
    model: { provider: "google", name: "gemini-2.5-pro", temperature: 0, max_tokens: null, api_key: null },
    system_prompt: "",
    tools: [],
    mcp: [],
    skills: [],
    memory: [],
    subagents: [],
    middleware: [],
    recursion_limit: 50,
    metadata: {},
  };
}

File: stream-backend/app/assistant_api.py
"""REST API for creating and managing folder-backed assistants.

Mounted by :mod:`app.main` via ``app.include_router(assistant_router)``. Every
assistant is a deepagents agent built from its :class:`AssistantConfig`.
"""
from __future__ import annotations

import logging
from typing import Any

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from .assistant_assist import generate_skill, generate_system_prompt, probe_model
from .assistant_catalog import catalog
from .bedrock_catalog import list_bedrock_models
from .assistants import (
    AssistantConfig,
    AssistantExists,
    AssistantNotFound,
    AssistantStore,
    MCPServerConfig,
    MemoryConfig,
    MiddlewareConfig,
    ModelConfig,
    SkillConfig,
    SubAgentConfig,
    ToolConfig,
    slugify,
)

logger = logging.getLogger("stream_backend.assistant_api")

router = APIRouter(prefix="/assistants", tags=["assistants"])

_store = AssistantStore()
try:
    _store.ensure_seeded()
except Exception:  # pragma: no cover - defensive
    logger.warning("assistant_api.seed_failed", exc_info=True)


def store() -> AssistantStore:
    return _store


# --------------------------------------------------------------------------
# Request/response bodies
# --------------------------------------------------------------------------
class AssistantUpsert(BaseModel):
    """Create/update payload. ``assistant_id`` is optional on create."""

    assistant_id: str | None = None
    name: str
    description: str = ""
    model: ModelConfig = Field(default_factory=ModelConfig)
    models: list[ModelConfig] = Field(default_factory=list)
    system_prompt: str = ""
    tools: list[ToolConfig] = Field(default_factory=list)
    mcp: list[MCPServerConfig] = Field(default_factory=list)
    skills: list[SkillConfig] = Field(default_factory=list)
    memory: list[MemoryConfig] = Field(default_factory=list)
    subagents: list[SubAgentConfig] = Field(default_factory=list)
    middleware: list[MiddlewareConfig] = Field(default_factory=list)
    recursion_limit: int = 50
    metadata: dict[str, Any] = Field(default_factory=dict)

    def to_config(self, assistant_id: str) -> AssistantConfig:
        data = self.model_dump()
        data["assistant_id"] = assistant_id
        return AssistantConfig.model_validate(data)


class SkillDraftRequest(BaseModel):
    name: str
    description: str = ""
    instructions: str = ""
    model: ModelConfig | None = None


class PromptDraftRequest(BaseModel):
    name: str
    description: str = ""
    tools: list[str] = Field(default_factory=list)
    subagents: list[str] = Field(default_factory=list)
    instructions: str = ""
    model: ModelConfig | None = None


class TestModelRequest(BaseModel):
    model: ModelConfig


class SkillWriteRequest(BaseModel):
    name: str
    content: str
    description: str = ""


class MemoryWriteRequest(BaseModel):
    name: str
    content: str


# --------------------------------------------------------------------------
# Catalog + AI-assist (declared before /{assistant_id} so they never collide)
# --------------------------------------------------------------------------
@router.get("/catalog")
async def get_catalog() -> dict[str, Any]:
    return catalog()


@router.get("/catalog/bedrock-models")
def get_bedrock_models() -> dict[str, Any]:
    # Sync handler: FastAPI runs the blocking boto3 calls in a threadpool.
    return list_bedrock_models()


@router.post("/assist/system-prompt")
async def assist_system_prompt(request: PromptDraftRequest) -> dict[str, Any]:
    logger.info("assistant_api.assist.system_prompt name=%s", request.name)
    return generate_system_prompt(
        name=request.name,
        description=request.description,
        tools=request.tools,
        subagents=request.subagents,
        model_config=request.model,
        instructions=request.instructions,
    )


@router.post("/assist/skill")
async def assist_skill(request: SkillDraftRequest) -> dict[str, Any]:
    logger.info("assistant_api.assist.skill name=%s", request.name)
    return generate_skill(
        name=request.name,
        description=request.description,
        model_config=request.model,
        instructions=request.instructions,
    )


@router.post("/assist/test-model")
def test_model(request: TestModelRequest) -> dict[str, Any]:
    # Sync handler: FastAPI runs it in a threadpool, so the blocking model
    # round-trip doesn't stall the event loop.
    logger.info(
        "assistant_api.assist.test_model provider=%s model=%s",
        request.model.provider,
        request.model.name,
    )
    return probe_model(request.model)


# --------------------------------------------------------------------------
# CRUD
# --------------------------------------------------------------------------
@router.get("")
async def list_assistants() -> list[dict[str, Any]]:
    return [config.model_dump(mode="json") for config in store().list()]


@router.post("", status_code=201)
async def create_assistant(payload: AssistantUpsert) -> dict[str, Any]:
    assistant_id = payload.assistant_id or slugify(payload.name)
    config = payload.to_config(assistant_id)
    try:
        created = store().create(config)
    except AssistantExists as exc:
        raise HTTPException(status_code=409, detail=f"Assistant '{assistant_id}' already exists") from exc
    logger.info("assistant_api.create id=%s", assistant_id)
    return created.model_dump(mode="json")


@router.get("/{assistant_id}")
async def get_assistant(assistant_id: str) -> dict[str, Any]:
    try:
        return store().get(assistant_id).model_dump(mode="json")
    except AssistantNotFound as exc:
        raise HTTPException(status_code=404, detail=f"Assistant '{assistant_id}' not found") from exc


@router.put("/{assistant_id}")
async def update_assistant(assistant_id: str, payload: AssistantUpsert) -> dict[str, Any]:
    config = payload.to_config(assistant_id)
    try:
        updated = store().update(assistant_id, config)
    except AssistantNotFound as exc:
        raise HTTPException(status_code=404, detail=f"Assistant '{assistant_id}' not found") from exc
    logger.info("assistant_api.update id=%s", assistant_id)
    return updated.model_dump(mode="json")


@router.delete("/{assistant_id}", status_code=204)
async def delete_assistant(assistant_id: str) -> None:
    try:
        store().delete(assistant_id)
    except AssistantNotFound as exc:
        raise HTTPException(status_code=404, detail=f"Assistant '{assistant_id}' not found") from exc
    logger.info("assistant_api.delete id=%s", assistant_id)


@router.post("/{assistant_id}/skills")
async def write_skill(assistant_id: str, request: SkillWriteRequest) -> dict[str, Any]:
    try:
        entry = store().write_skill(assistant_id, request.name, request.content, request.description)
    except AssistantNotFound as exc:
        raise HTTPException(status_code=404, detail=f"Assistant '{assistant_id}' not found") from exc
    return entry.model_dump(mode="json")


@router.post("/{assistant_id}/memory")
async def write_memory(assistant_id: str, request: MemoryWriteRequest) -> dict[str, Any]:
    try:
        entry = store().write_memory(assistant_id, request.name, request.content)
    except AssistantNotFound as exc:
        raise HTTPException(status_code=404, detail=f"Assistant '{assistant_id}' not found") from exc
    return entry.model_dump(mode="json")
